core/src/Kokkos_Pair.hpp

```diff
@@ -157,6 +157,60 @@ struct pair {
   std::pair<T1, T2> to_std_pair() const {
     return std::make_pair(first, second);
   }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION constexpr
+      typename std::enable_if<I == 0, T1 const&>::type
+      get() const noexcept {
+    return first;
+  }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION constexpr
+      typename std::enable_if<I == 1, T2 const&>::type
+      get() const noexcept {
+    return second;
+  }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION KOKKOS_CONSTEXPR_14
+      typename std::enable_if<I == 0, T1&>::type
+      get() noexcept {
+    return first;
+  }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION KOKKOS_CONSTEXPR_14
+      typename std::enable_if<I == 1, T2&>::type
+      get() noexcept {
+    return second;
+  }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION typename std::enable_if<I == 0, T1 volatile&>::type
+  get() volatile noexcept {
+    return first;
+  }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION typename std::enable_if<I == 1, T2 volatile&>::type
+  get() volatile noexcept {
+    return second;
+  }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION
+      typename std::enable_if<I == 0, T1 const volatile&>::type
+      get() const volatile noexcept {
+    return first;
+  }
+
+  template <size_t I>
+  KOKKOS_INLINE_FUNCTION
+      typename std::enable_if<I == 1, T2 const volatile&>::type
+      get() const volatile noexcept {
+    return second;
+  }
```

Comment:
Among other things, so that `Kokkos::pair` works with structured bindings, for instance. This is the standard-sanctioned way for interacting with `std::get` and all of the special magic that comes with that. The reason it has to be a member function is because you're not allowed to overload `std::get` directly.

What is your rationale for being so strongly against this?

Reply:
> What is your rationale for being so strongly against this?

* Does not exist in `std::pair`.
* Would require "dot-template" in generic code.

Reply:
The alternative is that we would have to provide a namespace-scope `get` that can be found via ADL (which I'd prefer to avoid exposing ADL interfaces to our users, hence the `Kokkos::get` Niebloid).  I think we can agree that `Kokkos::pair` should work with structured bindings, so those are the only two options.  The reason that `std::pair` does not do this is that it's in the same namespace as `std::get` already and can customize it directly.

Reply:
Also, users shouldn't use `val.template get<N>` in generic code; they should use `Kokkos::get<N>(val)`, just like the standard recommends w.r.t. `std::get`.

Reply:
for pair they should use val.first ;-)

Reply:
> The reason it has to be a member function is because you're not allowed to overload std::get directly.

Sure but we can also to specialize `Kokkos::get` for `Kokkos::Pair`.